pkg/types/rfc3161/v0.0.1/entry.go

```diff
@@ -0,0 +1,189 @@
+//
+// Copyright 2021 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rfc3161
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/asn1"
+	"encoding/base64"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/sigstore/rekor/pkg/types/rfc3161"
+
+	"github.com/go-openapi/strfmt"
+	"github.com/go-openapi/swag"
+	"github.com/sassoftware/relic/lib/pkcs9"
+	"github.com/sigstore/rekor/pkg/generated/models"
+	"github.com/sigstore/rekor/pkg/log"
+	"github.com/sigstore/rekor/pkg/types"
+)
+
+const (
+	APIVERSION = "0.0.1"
+)
+
+func init() {
+	if err := rfc3161.VersionMap.SetEntryFactory(APIVERSION, NewEntry); err != nil {
+		log.Logger.Panic(err)
+	}
+}
+
+type V001Entry struct {
+	Rfc3161Obj models.Rfc3161V001Schema
+}
+
+func (v V001Entry) APIVersion() string {
+	return APIVERSION
+}
+
+func NewEntry() types.EntryImpl {
+	return &V001Entry{}
+}
+
+func (v V001Entry) IndexKeys() []string {
+	var result []string
+
+	if v.HasExternalEntities() {
+		if err := v.FetchExternalEntities(context.Background()); err != nil {
+			log.Logger.Error(err)
+			return result
+		}
+	}
+
+	str := v.Rfc3161Obj.Tsr.Content.String()
+	tb, err := base64.StdEncoding.DecodeString(str)
+	if err != nil {
+		log.Logger.Warn(err)
+	} else {
+		h := sha256.Sum256(tb)
+		hx := hex.EncodeToString(h[:])
+
+		payloadKey := "sha256:" + hx
+		result = append(result, payloadKey)
+	}
+
+	return result
+}
+
+func (v *V001Entry) Unmarshal(pe models.ProposedEntry) error {
+	rfc3161Resp, ok := pe.(*models.Rfc3161)
+	if !ok {
+		return errors.New("cannot unmarshal non Rfc3161 v0.0.1 type")
+	}
+
+	if err := types.DecodeEntry(rfc3161Resp.Spec, &v.Rfc3161Obj); err != nil {
+		return err
+	}
+
+	// field validation
+	if err := v.Rfc3161Obj.Validate(strfmt.Default); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (v V001Entry) HasExternalEntities() bool {
+	return false
+}
+
+func (v *V001Entry) FetchExternalEntities(ctx context.Context) error {
+	return nil
+}
+
+func (v *V001Entry) Canonicalize(ctx context.Context) ([]byte, error) {
+	if err := v.FetchExternalEntities(ctx); err != nil {
+		return nil, err
+	}
+
+	if v.Rfc3161Obj.Tsr == nil || v.Rfc3161Obj.Tsr.Content == nil {
+		return nil, fmt.Errorf("tsr content is nil")
+	}
+
+	content := *v.Rfc3161Obj.Tsr.Content
+
+	// validate the content
```

Comment:
should we move this validation logic into `Validate` and store the cert in a struct field?

Reply:
Looking through the code (I'm new), it doesn't seem like validate is called before "writing" an entry to log? I only see it in SearchLogQueryHandler? Am I missing something?

Reply:
In RPM/JAR/Rekord, we call it in `FetchExternalEntities`; in in-toto it's called at the end of `Unmarshal`; perhaps it was an errant delete since `FetchExternalEntities` is a no-op in this type?

Reply:
Yeah the combined typing for sending data and storing is little confusing. This type probably doesn't need to fetch external entries, but I could do validation in `Unmarshal`.

Reply:
I also find the model of types exporting all their methods and then reusing them internally to be somewhat confusing. Is an external system orchestrating all the actions or is the struct supposed to do everything itself? It really doesn't help define what the scope of responsibility is.

Reply:
Anyway, I think was able to follow the current convention and store only hash+cert. Lemme know if that is sufficient? Or if we should extract a signature as well.